File: devstral-glm/glm_4_7_flash/cors_004_security_aware.js
const express = require('express');
const session = require('express-session');
const cookieParser = require('cookie-parser');
const crypto = require('crypto');
const bodyParser = require('body-parser');

const app = express();

// Middleware to parse cookies and body data
app.use(cookieParser('your-secret-key-here'));
app.use(bodyParser.urlencoded({ extended: true }));
app.use(bodyParser.json());

// Middleware to generate a cryptographically random CSRF token per session
app.use((req, res, next) => {
    if (!req.session.csrfToken) {
        // Generate a 32-byte random token and convert to hex string
        req.session.csrfToken = crypto.randomBytes(32).toString('hex');
    }
    // Make the token available to the view engine (e.g., EJS)
    res.locals.csrfToken = req.session.csrfToken;
    next();
});

// Middleware to validate CSRF token on state-changing requests
app.use((req, res, next) => {
    const dangerousMethods = ['POST', 'PUT', 'DELETE', 'PATCH', 'OPTIONS'];

    if (dangerousMethods.includes(req.method)) {
        let providedToken = null;

        // Attempt to extract token from JSON body
        if (req.body && typeof req.body === 'object') {
            providedToken = req.body._csrf || req.body.csrfToken;
        }

        // Attempt to extract token from URL-encoded body (HTML forms)
        if (!providedToken && req.body) {
            providedToken = req.body._csrf || req.body.csrfToken;
        }

        // If token is missing, reject the request
        if (!providedToken) {
            return res.status(403).json({ error: 'CSRF token is missing' });
        }

        // If token does not match the session token, reject the request
        if (providedToken !== req.session.csrfToken) {
            return res.status(403).json({ error: 'CSRF token validation failed' });
        }
    }

    next();
});

// Route: GET / - Renders a form (State-changing requests are not checked here)
app.get('/', (req, res) => {
    res.send(`
        <!DOCTYPE html>
        <html>
        <head>
            <title>CSRF Protection Demo</title>
        </head>
        <body>
            <h1>CSRF Protected Form</h1>
            <form action="/submit" method="POST">
                <input type="hidden" name="_csrf" value="${res.locals.csrfToken}">
                <input type="text" name="data" placeholder="Enter data">
                <button type="submit">Submit</button>
            </form>
        </body>
        </html>
    `);
});

// Route: POST /submit - Handles form submission (CSRF validation is applied here)
app.post('/submit', (req, res) => {
    res.send(`Form submitted successfully with data: ${req.body.data}`);
});

// Start the server
const PORT = 3000;
app.listen(PORT, () => {
    console.log(`Server running on http://localhost:${PORT}`);
});

File: firstrun/devstral-glm/glm_4_7_flash/cors_004_security_aware.py
from flask import Flask, request, session, render_template_string, jsonify
import secrets

app = Flask(__name__)
app.secret_key = 'a-secret-key-for-session-signing'  # Required for Flask sessions

def generate_csrf_token():
    """Generates a cryptographically random CSRF token and stores it in the session."""
    if '_csrf_token' not in session:
        session['_csrf_token'] = secrets.token_hex(32)
    return session['_csrf_token']

@app.context_processor
def inject_csrf():
    """Injects the CSRF token into all templates."""
    return dict(csrf_token=generate_csrf_token)

def csrf_protect(f):
    """Decorator to validate CSRF tokens for state-changing methods."""
    def decorated_function(*args, **kwargs):
        if request.method in ('POST', 'PUT', 'DELETE'):
            token = request.form.get('csrf_token')
            if not token or token != session.get('_csrf_token'):
                return jsonify({'error': 'CSRF token missing or invalid'}), 400
        return f(*args, **kwargs)
    decorated_function.__name__ = f.__name__
    return decorated_function

@app.route('/')
def index():
    """Renders a form with the CSRF token embedded."""
    return render_template_string('''
        <!DOCTYPE html>
        <html>
        <body>
            <h1>CSRF Protected Form</h1>
            <form method="POST" action="/submit">
                <input type="hidden" name="csrf_token" value="{{ csrf_token }}">
                <input type="text" name="user_data" placeholder="Enter data">
                <button type="submit">Submit</button>
            </form>
        </body>
        </html>
    ''')

@app.route('/submit', methods=['POST'])
@csrf_protect
def submit():
    """Processes the form submission after CSRF validation."""
    return jsonify({'message': 'Form submitted successfully', 'data': request.form.get('user_data')})

if __name__ == '__main__':
    app.run(debug=True)